packages/flutter/lib/src/gestures/pointer_signal_arena.dart

```diff
@@ -0,0 +1,90 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'binding.dart';
+import 'events.dart';
+
+/// The callback for a winning [PointerSignalArenaMember]
+typedef PointerSignalResolvedCallback = void Function(PointerEvent event);
+
+/// Represents an object participating in a pointer signal arena.
+///
+/// An object interested in handling a pointer signal event should create
+/// an instance of this class to enter itself in the arena for that event. The
+/// provided callback will be called if it successfully claims the event.
+class PointerSignalArenaMember {
+  /// Creates a new arena member, which immediately enters itself in the
+  /// pointerSignalArena.
+  PointerSignalArenaMember(this.event, this.acceptCallback) {
+    GestureBinding.instance.pointerSignalArena.add(event, this);
+  }
+
+  /// The event for the pointer signal.
+  PointerEvent event;
+
+  /// The callback to call if [acceptSignal] is called.
+  PointerSignalResolvedCallback acceptCallback;
+
+  /// Called when this member wins the arena for the given pointer event.
+  void acceptSignal() {
+    acceptCallback(event);
+  }
+}
+
+class _PointerSignalArena {
+  final List<PointerSignalArenaMember> members = <PointerSignalArenaMember>[];
+  bool isOpen = true;
+
+  void add(PointerSignalArenaMember member) {
+    assert(isOpen);
+    members.add(member);
+  }
+
+  @override
+  String toString() {
+    final StringBuffer buffer = StringBuffer();
+    if (members.isEmpty) {
+      buffer.write('<empty>');
+    } else {
+      buffer.write(members.join(', '));
+    }
+    if (isOpen) {
+      buffer.write(' [open]');
+    }
+    return buffer.toString();
+  }
+}
+
+/// An arena for use with pointer signal events.
+///
+/// Pointer signals are immediate, so unlike a gesture arena it always resolves
+/// when the arena closes.
+class PointerSignalArenaManager {
+  final Map<PointerSignalEvent, _PointerSignalArena> _arenas =
+      <PointerSignalEvent, _PointerSignalArena>{};
+
+  /// Adds a new member to the arena.
+  void add(PointerSignalEvent event, PointerSignalArenaMember member) {
+    final _PointerSignalArena state = _arenas.putIfAbsent(event, () {
+      return _PointerSignalArena();
+    });
+    state.add(member);
+  }
+
+  /// Prevents new members from entering the arena, and resolves it.
+  ///
+  /// Called after the framework has finished dispatching the pointer signal
+  /// event.
+  ///
+  /// Currently, the resolution is always that the first entry in the arena
+  /// wins.
+  void close(PointerSignalEvent event) {
+    final _PointerSignalArena state = _arenas[event];
+    if (state == null)
+      return; // This arena either never existed or has been resolved.
+    state.isOpen = false;
+    state.members.first.acceptSignal();
+    _arenas.remove(event);
+  }
+}
```

Comment:
i'm confused as to why this has to be this complicated. Can you discuss this design?

Reply:
- `PointerSignalArenaMember` exists because there needs to be a way to register a callback on winning, and this was consistent with the arena structure so I thought it would be more familiar (and extensible, in case, e.g., something needs a `rejectCallback` for one of these events in the future). I could just have clients register `PointerSignalResolvedCallback`s directly in the list if you'd prefer, but that seems less future-proof.
- The `Arena`/`ArenaManager` split isn't actually necessary given that these are resolved immediately; it shouldn't be possible at the moment to have more than one active event at a time here. I kept that aspect of the gesture arena structure partially speculatively, and partially because it avoids any possibility of members someone registering in the same arena but for two different events. I could simplify to a single arena, possibly with an assert on the event being processed, if you'd prefer.

Does that cover your concerns, or are there other parts that you think are too complex?

Reply:
I think it's interesting that you kept the arena/arena manager split (which i would have folded into one thing) but simplified from a member interface to a callback (which i would have kept as an interface). :-)

I think we could simplify further. There's no need to actually keep the list, since we only care about the first thing in the list. I'm not sure calling this an "arena" is really getting us much either. We could just call it a "PointerSignalDispatcher" or "PointerSignalTiebreaker" or something like that. Its role is just to decide who gets the signal.

Can you talk more about the case where we can imagine having two signal events being disambiguated at the same time?

Reply:
> I think it's interesting that you kept the arena/arena manager split

I didn't; I didn't remove manager from the name which may have caused confusion?

> but simplified from a member interface to a callback (which i would have kept as an interface). :-)

I'll add that back since per my other comment I think that's worth keeping add well.

> I think we could simplify further. There's no need to actually keep the list, since we only care about the first thing in the list.

True, since it's internal the list could be added later if we need it at some point. Will update.

> We could just call it a "PointerSignalDispatcher" or "PointerSignalTiebreaker" or something like that.

Sure. I assumed since in original discussions you called it an arena you thought it was similar in concept so I kept that term rather than introducing a new one, but I'm not attached to the name. Will update in the next round.

> Can you talk more about the case where we can imagine having two signal events being disambiguated at the same time?

This version doesn't handle that, it just asserts that it hasn't happened somehow.

Reply:
I was going from your description [above](https://github.com/flutter/flutter/pull/22762/files#r256186584), I see what you mean in the code now.

I'm not suggesting you should have the interface. A callback is probably fine too. I think we should just handle the cases we know of, especially if we don't have any specific cases in mind that we know we want to handle one day.

Reply:
Simplified further per your comments. I left the callback as it was in the previous iteration to avoid thrashing, since I'm confused at this point as to whether you prefer the direct callback or the member with a callback parameter. I'm happy to add it back given clear guidance.

The arena has been renamed to `PointerSignalResolver`. "Dispatcher" seemed potentially confusing since the binding layer refers to the earlier stage of delivering the events to hit-tested target as "dispatching". "Tiebreaker" could suggest that there's an actual tie, rather than a clear precedence (leaf-most node interested in the event should win).